### ISSUE TITLE
SQLFluff adds whitespace before Typed Date Literal 

### ISSUE THREAD

**Issue by R7L208** [sqlfluff/sqlfluff#2289]
```sql
INSERT INTO students PARTITION (birthday = DATE'2019-01-02')
VALUES ('Amy Smith', '123 Park Ave, San Jose');
```

## Expected Behaviour
Lint and Fix should not flag that an extra whitespace is needed between `DATE` and the string literal (`'2019-01-02'`)
## Observed Behaviour
fix changes the code to the below which breaks the typed date literal

```sql
INSERT INTO students PARTITION (birthday = DATE '2019-01-02')
VALUES ('Amy Smith', '123 Park Ave, San Jose');
```
## Steps to Reproduce
Run SQLFluff lint/fix on code segment provided above.

## Dialect
spark3

## Version
Include the output of `sqlfluff --version` along with your Python version.
If you are using dbt then please additionally include the dbt version.
sqlfluff, version 0.9.1

## Configuration
```
==== sqlfluff ====
sqlfluff:                0.9.1 python:                 3.10.1
implementation:        cpython dialect:                spark3
verbosity:                   3 templater:               jinja

== Raw Config:
core:
    dialect:            spark3
    disable_noqa:       False
    encoding:           autodetect
    ignore:             []
    ignore_templated_areas:True
    nocolor:            False
    output_line_length: 80
    recurse:            True
    runaway_limit:      10
    sql_file_exts:      .sql,.sql.j2,.dml,.ddl
    templater:          jinja
    verbose:            3
indentation:
    indented_joins:     False
    indented_using_on:  True
    template_blocks_indent:True
rules:
    allow_scalar:       True
    comma_style:        trailing
    indent_unit:        space
    max_line_length:    80
    single_table_references:consistent
    tab_space_size:     4
    unquoted_identifiers_policy:all
    Example_L001:
        forbidden_columns:  bar, baaz
    L007:
        operator_new_lines: after
    L010:
        capitalisation_policy:consistent
    L011:
        aliasing:           explicit
    L012:
        aliasing:           explicit
    L014:
        extended_capitalisation_policy:consistent
    L016:
        ignore_comment_lines:False
    L026:
        force_enable:       False
    L028:
        force_enable:       False
    L029:
        unquoted_identifiers_policy:aliases
    L030:
        capitalisation_policy:consistent
    L038:
        select_clause_trailing_comma:forbid
    L040:
        capitalisation_policy:consistent
    L042:
        forbid_subquery_in: join
    L047:
        prefer_count_0:     False
        prefer_count_1:     False
    L052:
        multiline_newline:  False
        require_final_semicolon:False
    L054:
        group_by_and_order_by_style:consistent
    L057:
        additional_allowed_characters:""
        allow_space_in_identifier:False
        quoted_identifiers_policy:all
        unquoted_identifiers_policy:all
templater:
    unwrap_wrapped_queries:True
    jinja:
        apply_dbt_builtins: True
        macros:
            dbt_config:         {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
            dbt_is_incremental: {% macro is_incremental() %}True{% endmacro %}
            dbt_ref:            {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
            dbt_source:         {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
            dbt_var:            {% macro var(variable, default='') %}item{% endmacro %}
```

**Comment by R7L208** [comment]
Will update Steps to Reproduce but for now code is in draft pr #2290

**Comment by barrywhart** [comment]
Are you saying that the bug only occurs with new code that's in a PR?

**Comment by R7L208** [comment]
No, this occurs even if PR is not merged but a test case to replicate is here.

**Comment by barrywhart** [comment]
Using your test SQL, I'm getting a parse error. 🤔
```
(sqlfluff-3.10.1) ➜  sqlfluff git:(main) ✗ sqlfluff lint test.sql --dialect spark3
== [test.sql] FAIL                                                                                                                                                                                      
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'INSERT
                       | INTO students PARTITION (birthday...'
All Finished 📜 🎉!
```

**Comment by jpy-git** [comment]
@R7L208 So this is valid syntax in Snowflake (and I think postgres too):

![image](https://user-images.githubusercontent.com/80432516/149764886-b38515ca-22f5-4f69-9ccb-d3de9d2288f0.png)

Looking at the Spark3 [docs](https://spark.apache.org/docs/latest/sql-ref-literals.html#datetime-literal) it actually looks like they use spaces between DATE and the date string in their examples so seems like the rule is arguably currently improving the formatting:

![image](https://user-images.githubusercontent.com/80432516/149765349-d9ba87dd-1166-4452-8d36-8aec71576dc3.png)

> fix changes the code to the below which breaks the typed date literal

Given the doc examples, does this fix actually `break` the spark3 syntax in the sense that it results in unparsable code, or is it personal preference to not have it split?

**Comment by R7L208** [comment]
Ah, thank you. I missed that in the docs and have only seen it without a space.

Would be a personal preference then but do actually prefer that it follows syntax as in the docs as implemented now, so will be closing the issue
